src/sources/nats.rs

```diff
@@ -0,0 +1,181 @@
+use crate::{
+    config::{log_schema, DataType, SourceConfig, SourceContext, SourceDescription},
+    event::{Event, Value},
+    internal_events::NatsEventReceived,
+    shutdown::ShutdownSignal,
+    Pipeline,
+};
+use bytes::Bytes;
+use chrono::Utc;
+use futures::{pin_mut, stream, SinkExt, Stream, StreamExt};
+use serde::{Deserialize, Serialize};
+use snafu::Snafu;
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Could not create Nats subscriber: {}", source))]
+    NatsCreateError { source: std::io::Error },
+    #[snafu(display("Could not subscribe to Nats topics: {}", source))]
+    NatsSubscribeError { source: std::io::Error },
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct NatsSourceConfig {
+    url: String,
+    connection_name: String,
+    subject: String,
+    queue: Option<String>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<NatsSourceConfig>("nats")
+}
+
+impl_generate_config_from_default!(NatsSourceConfig);
```

Comment:
I think the generated config should more closely match what the `nats` sink generates. Currently it generates:

```toml
[sources.source0]
connection_name = ""
subject = ""
type = "nats"
url = ""
```

Where I think:

```toml
[sources.source0]
connection_name = "vector"
subject = "from.vector"
type = "nats"
url = "nats://127.0.0.1:4222"
```

Would be better.

You can see it uses a static block for the generated config: https://github.com/timberio/vector/blob/23305a2a827370978471643751c62792ab2666a9/src/sinks/nats.rs#L55-L61

But you could also define the `Default` trait here, rather than deriving it, and continue using `impl_generate_config_from_default`.